DR-2907: Autoclass Managed Buckets

snf2ye opened the pull request:
<img width="923" alt="image" src="https://user-images.githubusercontent.com/13254229/220513201-ddf97998-3da3-4f4c-bfa5-689408f373f2.png">

Included Changes:
- Added autoclass_enabled boolean field on bucket_resource table
- On Bucket create for file ingests, autoclass will be turned on by default and this setting will be recorded in the bucket_resource table

More information on autoclass can be found in [this spike document](https://docs.google.com/document/d/1j7M6cMHgrC0V6gITxdIBSPgDjwXcyUjuG9n83mrr7nk/edit) and in the [google documentation](https://cloud.google.com/storage/docs/autoclass). 

Disclaimer: Apologies for the mess of the commits. They are not meaningful.

snf2ye committed:
Changes from BOM PR

snf2ye committed:
First pass - set autoclass enabled at appropriate locations; upgrade libraries

spotless

switch to boolean

snf2ye committed:
Fix unit tests

Fix tests

snf2ye committed:
v2 for everything

snf2ye committed:
rework ondemand test

on demand

snf2ye committed:
Add tests

snf2ye committed:
proved with test that we can set multiregional bucket and have autoclass

snf2ye committed:
remove comment

snf2ye committed:
spotless & test spotbugs

nmalfroy commented on src/main/java/bio/terra/service/resourcemanagement/ResourceService.java:
```diff
@@ -161,7 +161,8 @@ public GoogleBucketResource getOrCreateBucketForFile(
         getReaderGroups,
-        dedicatedServiceAccount);
+        dedicatedServiceAccount,
+        true);
```
shouldn't this be based on the dataset's option?

snf2ye replied:
(I think this is covered with the other thread)

nmalfroy commented on src/main/java/bio/terra/service/resourcemanagement/ResourceService.java:
```diff
@@ -187,7 +188,8 @@ public GoogleBucketResource getOrCreateBucketForBigQueryScratchFile(
         null,
-        dataset.getProjectResource().getServiceAccount());
+        dataset.getProjectResource().getServiceAccount(),
+        false);
```
Might be worth a comment in the method's javadoc that this is always false because autoclass isn't worth the cost given the scratch bucket's usage pattern?

nmalfroy commented on src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java:
```diff
@@ -235,12 +242,13 @@ private GoogleBucketResource createMetadataRecord(
     // insert a new bucket_resource row and lock it
     GoogleBucketResource googleBucketResource =
-        resourceDao.createAndLockBucket(bucketName, projectResource, region, flightId);
+        resourceDao.createAndLockBucket(bucketName, projectResource, region, flightId, true);
```
shouldn't this be
```suggestion
        resourceDao.createAndLockBucket(bucketName, projectResource, region, flightId, autoclassEnabled);
```

snf2ye replied:
Yes, you're totally right. Thank you for catching this.
I think the existence of this error shows that I don't have sufficient test coverage. I will add some alongside this fix.

nmalfroy commented on src/main/java/bio/terra/service/resourcemanagement/google/GoogleBucketService.java:
```diff
@@ -356,13 +377,10 @@ private Bucket newCloudBucket(
             // .setRequesterPays()
-            // See here for possible values: http://g.co/cloud/storage/docs/storage-classes
-            .setStorageClass(storageClass)
+            .setAutoclass(Autoclass.newBuilder().setEnabled(enableAutoclass).build())
```
oh man, they made the API kinda weird

nmalfroy reviewed (changes requested):
looks good. I just had the one question about the place where you're setting autoclass to true instead of using the bucket resource's value (I may be misunderstanding). Also, just wanted to confirm. did we say we weren't exposing the option to set autoclass in the API?

snf2ye commented:
> Also, just wanted to confirm. did we say we weren't exposing the option to set autoclass in the API?

@nmalfroy - We decided we didn't want to expose this option. It's low enough overhead to the user that we are turning it on by default.

okotsopoulos commented:
> @nmalfroy - We decided we didn't want to expose this option. It's low enough overhead to the user that we are turning it on by default.

@snf2ye I thought that we wanted to enable this by default but allow a user to opt out of it if desired? Maybe I'm remembering incorrectly.

snf2ye commented:
> > @nmalfroy - We decided we didn't want to expose this option. It's low enough overhead to the user that we are turning it on by default.
> 
> I thought that we wanted to enable this by default but allow a user to opt out of it if desired? Maybe I'm remembering incorrectly.

I believe we said that it was only a super user that would request autoclass to be turned off and that we could handle these manually in a one-off manner. 

The work completed in this PR matches the scope defined in [the ticket](https://broadworkbench.atlassian.net/browse/DR-2907).

snf2ye committed:
PR Review: Add test; Fix variable; add javadoc note

- I didn't want to add a full test that creates a whole project/bucket/etc in order to test these scenarios so I'm piggy backing on related tests.
- The tests now check for the autoclass setting in two places: postgres bucket_resource setting and bucket info returned about the cloud bucket
- Added note about why autoclass class is turned on/off for each bucket type
- Fix erroneous "true" setting and now passing through autoclassEnabled variable in GoogleBucketService

nmalfroy reviewed (approved):
great!
